test/e2e/app/app.go

```diff
@@ -296,66 +304,74 @@ func (app *Application) ApplySnapshotChunk(_ context.Context, req *abci.RequestA
 // construct a special transaction whose value is the sum of all of the vote
 // extensions from the previous round.
 //
-// NB: Assumes that the supplied transactions do not exceed `req.MaxTxBytes`.
+// Additionally, we verify the vote extension signatures passed from CometBFT and
+// include all data necessary for such verification in the special transaction's
+// payload so that ProcessProposal at other nodes can also verify the proposer
+// constructed the special transaction correctly
+//
 // If adding a special vote extension-generated transaction would cause the
 // total number of transaction bytes to exceed `req.MaxTxBytes`, we will not
 // append our special vote extension transaction.
 func (app *Application) PrepareProposal(
 	_ context.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
-	var sum int64
-	var extCount int
-	for _, vote := range req.LocalLastCommit.Votes {
-		if !vote.SignedLastBlock || len(vote.VoteExtension) == 0 {
-			continue
-		}
-		extValue, err := parseVoteExtension(vote.VoteExtension)
-		// This should have been verified in VerifyVoteExtension
-		if err != nil {
-			panic(fmt.Errorf("failed to parse vote extension in PrepareProposal: %w", err))
-		}
-		valAddr := crypto.Address(vote.Validator.Address)
-		app.logger.Info("got vote extension value in PrepareProposal", "valAddr", valAddr, "value", extValue)
-		sum += extValue
-		extCount++
+
+	extCount, sum, err := app.verifyAndSum(req.Height, &req.LocalLastCommit, "prepareproposal")
+	if err != nil {
+		panic(fmt.Errorf("failed to sum and verify in PrepareProposal; err %w", err))
 	}
 
 	if app.cfg.PrepareProposalDelay != 0 {
 		time.Sleep(app.cfg.PrepareProposalDelay)
 	}
 
 	// We only generate our special transaction if we have vote extensions
+	extTxPrefix := fmt.Sprintf("%s=", voteExtensionKey)
 	if extCount > 0 {
 		var totalBytes int64
-		extTxPrefix := fmt.Sprintf("%s=", voteExtensionKey)
-		extTx := []byte(fmt.Sprintf("%s%d", extTxPrefix, sum))
-		app.logger.Info("preparing proposal with custom transaction from vote extensions", "tx", extTx)
-		// Our generated transaction takes precedence over any supplied
-		// transaction that attempts to modify the "extensionSum" value.
-		txs := make([][]byte, len(req.Txs)+1)
-		for i, tx := range req.Txs {
+		extCommitBytes, err := req.LocalLastCommit.Marshal()
```

Comment:
As this serves as an example app, this whole loop could be applied irrespective of `extCount>0` to guard against transactions starting with `extTxPrefix` and `reservedKey`.
Then build and append the special transaction only if `extCount>0`.

Reply:
I like your suggestion, let me see what I can do.
Also, for the future, we can strengthen the logic of this function, because we do know when vote extensions are active and when they are not. The way it is now is a bit loose, and, since this is a testing app we should probable make if panic if we receive extensions when we shouldn't (or if they are missing when they shouldn't)

Reply:
Actually, I decided to go for the "tightening up" the checks we do on vote extensions. Everything is pretty optional in this PR. I have an upcoming PR lined up which does that, and simplifies this loop massively. Please take a look when it's up

Reply:
#509